lib/text_diff.ml

```diff
@@ -65,54 +80,86 @@ let process_type_diff (type_diff : Diff.type_) =
-        in
-        String_map.update module_path
-          (function
-            | None -> Some diff | Some existing -> Some (existing @ diff))
-          acc
-    | Modified (Supported changes) ->
-        List.fold_left
-          (fun acc' change ->
-            match (change : Diff.sig_item) with
-            | Value val_diff ->
-                let diff = process_value_diff val_diff in
-                String_map.update module_path
-                  (function
-                    | None -> Some diff | Some existing -> Some (existing @ diff))
-                  acc'
-            | Type type_diff ->
-                let diff = process_type_diff type_diff in
-                String_map.update module_path
-                  (function
-                    | None -> Some diff | Some existing -> Some (existing @ diff))
-                  acc'
-            | Module sub_module_diff ->
-                let sub_module_path =
-                  match sub_module_diff.mdiff with
-                  | Modified _ -> module_path ^ "." ^ sub_module_diff.mname
-                  | Added _ | Removed _ -> module_path
-                in
-                process_module_diff sub_module_path sub_module_diff acc')
-          acc changes
-  in
+let rec process_module_type_diff module_path (module_type_diff : Diff.modtype)
+    acc =
+  match module_type_diff.mtdiff with
+  | Added curr_mtd ->
+      let diff =
+        [ { orig = []; new_ = mtd_to_lines module_type_diff.mtname curr_mtd } ]
+      in
+      String_map.update module_path
+        (function None -> Some diff | Some existing -> Some (existing @ diff))
+        acc
+  | Removed ref_mtd ->
+      let diff =
+        [ { orig = mtd_to_lines module_type_diff.mtname ref_mtd; new_ = [] } ]
+      in
+      String_map.update module_path
+        (function None -> Some diff | Some existing -> Some (existing @ diff))
+        acc
+  | Modified Unsupported ->
+      String_map.add module_path
+        [ { orig = []; new_ = [ "<unsupported change>" ] } ]
+        acc
+  | Modified (Supported changes) ->
+      module_modification_changes module_path changes acc
+
+and process_module_diff module_path (module_diff : Diff.module_) acc =
+  match module_diff.mdiff with
+  | Modified Unsupported ->
+      String_map.add module_path
+        [ { orig = []; new_ = [ "<unsupported change>" ] } ]
+        acc
+  | Added curr_md ->
+      let diff =
+        [ { orig = []; new_ = md_to_lines module_diff.mname curr_md } ]
+      in
+      String_map.update module_path
+        (function None -> Some diff | Some existing -> Some (existing @ diff))
+        acc
+  | Removed ref_md ->
+      let diff =
+        [ { orig = md_to_lines module_diff.mname ref_md; new_ = [] } ]
+      in
+      String_map.update module_path
+        (function None -> Some diff | Some existing -> Some (existing @ diff))
+        acc
+  | Modified (Supported changes) ->
+      module_modification_changes module_path changes acc
```

Comment:
Those two functions (`process_module_type_diff` and `process_module_diff`) are still very similar.

Factoring them would be make the code much better I think. However, this could be done in another PR, not to delay this one.

If you want to try to factor these function, what you can do is:
- extract all the common part into a function
- fill the non-common part with variables, passed as input
- call the function from `process_module_diff` and `process_module_type_diff` with the right arguments
- Report any problem you (will) have ;)

Reply:
I've tried to refactor the code to avoid duplication. Is it in a better state now? Is there something I can add ?

Reply:
That is not exactly what I had in mind, but it is fine. Basically, you factored out everything, except the function `process_module_diff` and `process_module_type_diff` :wink:.

Here is what I had in mind:

```ocaml
let rec process_sig_diff :
    type a. _ -> (string -> a -> string list) -> (a, _) Diff.t * _ -> _ -> _ =
 fun path to_lines ((diff : (a, _) Diff.t), name) acc ->
  match diff with
  | Added curr_mtd ->
      let diff = [ { orig = []; new_ = to_lines name curr_mtd } ] in
      String_map.update path
        (function None -> Some diff | Some existing -> Some (existing @ diff))
        acc
  | Removed ref_mtd ->
      let diff = [ { orig = to_lines name ref_mtd; new_ = [] } ] in
      String_map.update path
        (function None -> Some diff | Some existing -> Some (existing @ diff))
        acc
  | Modified Diff.Unsupported ->
      String_map.add path
        [ { orig = []; new_ = [ "<unsupported change>" ] } ]
        acc
  | Modified (Supported changes) -> module_modification_changes path changes acc

and process_module_type_diff module_path (module_type_diff : Diff.modtype) acc =
  process_sig_diff module_path mtd_to_lines
    (module_type_diff.mtdiff, module_type_diff.mtname)
    acc

and process_module_diff module_path (module_diff : Diff.module_) acc =
  process_sig_diff module_path md_to_lines
    (module_diff.mdiff, module_diff.mname)
    acc

and module_modification_changes module_path items acc =
 [...]
```

the `type a . <type>` is needed because I call the function `process_sig_diff` recursively with different types. See [here](https://ocaml.org/manual/5.2/gadts-tutorial.html#s:gadts-recfun) for more information!

Your factoring of the printing function is good, but we should keep one printing function per type (`md_to_lines`, `mtd_to_lines`, ...) to avoid dealing with printing in the middle of some other logic (choosing with delimiter to use, ...).

Reply:
(That would be even more unrelated to this PR, but one could factor the `process_value_diff` and `process_type_diff` and divide by two the number of lines they take by removing the duplicated logic.)

Reply:
I tried to do this but kept having a tpye mismatch. Are you using a gadts? that's awesome

Reply:
Thank you. I just saw the link to the documentation you attached.